src/gui/cmdline.cpp

```diff
@@ -0,0 +1,118 @@
+#include "cmdline.h"
+#include "cmdwidget.h"
+
+namespace NeovimQt {
+
+CmdLine::CmdLine(QWidget *parent) : QTextEdit(parent) {
+        hide();
+
+        setFrameShape(QFrame::NoFrame);
+        setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+
+        document()->setDefaultFont(parent->font());
+
+        auto cmdwidget_parent = dynamic_cast<CmdWidget*>(parent);
+        setTextBackgroundColor(cmdwidget_parent->shell()->background());
+        setTextColor(cmdwidget_parent->shell()->foreground());
+}
+
+void CmdLine::compute_document(const QString& firstc, const QString& prompt, const QVariantList& content) {
+        QString plaintext_content;
+        QString html_content;
+        clear();
+        style_cursor_CharFormat(QVariantMap());
+        textCursor().insertText(firstc + prompt);
+        foreach(QVariant piece, content){
+                qDebug() << piece.toList();
+                style_cursor_CharFormat(piece.toList().at(0).toMap());
+                plaintext_content = piece.toStringList().at(1);
+                textCursor().insertText(plaintext_content, textCursor().charFormat());
+        }
+        setFocus();
+}
+
+QColor CmdLine::color(qint64 color, const QColor& fallback) const {
+        if (color == -1) {
+                return fallback;
+        }
+        return QRgb(color);
+}
+
+void CmdLine::style_cursor_CharFormat(const QVariantMap& attrs) {
+        QTextCharFormat cur_format;
+
+        auto cmdwidget_parent = dynamic_cast<CmdWidget*>(parent());
```

Comment:
Since we are dependent on the widget anyway - it is ok to store the proper pointer type in a member of this class and use it here, avoiding the dynamic_cast.

Reply:
or the shell directly

Reply:
No problem. I'll use the CmdWidget for this, it seems more logical to me.

For the time being `CmdWidget::shell()` exists, but maybe at some point we could want to use separate styling (basically separate `QPalette` I guess) for Shell and CmdWidget